File: android/src/main/java/com/library/PayjoinModule.java
package com.library;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Promise;
import org.bitcoinj.core.*;
import org.bitcoinj.params.TestNet3Params;
import org.bitcoinj.script.Script;
import org.bitcoinj.wallet.Wallet;

public class PayjoinModule extends ReactContextBaseJavaModule {
    private static final NetworkParameters PARAMS = TestNet3Params.get();

    public PayjoinModule(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @Override
    public String getName() {
        return "PayjoinModule";
    }

    @ReactMethod
    public void createInitialPsbt(String senderAddress, String receiverAddress, int amount, int feeRate, Promise promise) {
        try {
            Wallet wallet = new Wallet(PARAMS);
            Transaction tx = new Transaction(PARAMS);

            // Fetch UTXOs (simplified, assuming a helper method)
            List<Utxo> utxos = fetchUtxos(senderAddress);

            long totalInput = 0;
            for (Utxo utxo : utxos) {
                Sha256Hash txid = Sha256Hash.wrap(utxo.getTxid());
                tx.addInput(txid, utxo.getVout(), new Script(Hex.decode(utxo.getScriptPubKey())));
                totalInput += utxo.getAmount();
            }

            // Add receiver output
            Address receiverAddr = Address.fromString(PARAMS, receiverAddress);
            tx.addOutput(Coin.valueOf(amount), receiverAddr);

            // Add change output
            long estimatedFee = feeRate * 200 + 10000; // Max fee contribution
            long change = totalInput - amount - estimatedFee;
            if (change > 546) { // Dust threshold
                Address changeAddr = Address.fromString(PARAMS, senderAddress);
                tx.addOutput(Coin.valueOf(change), changeAddr);
            }

            // Convert to PSBT (simplified, assumes conversion utility)
            String psbtBase64 = convertToPsbt(tx);
            promise.resolve(psbtBase64);
        } catch (Exception e) {
            promise.reject("PSBT_CREATION_FAILED", e);
        }
    }

    @ReactMethod
    public void signAndFinalizePsbt(String psbtBase64, String privateKeyWif, Promise promise) {
        try {
            // Parse PSBT (simplified, assumes conversion utility)
            Transaction tx = convertFromPsbt(psbtBase64);

            // Sign transaction
            ECKey key = DumpedPrivateKey.fromBase58(PARAMS, privateKeyWif).getKey();
            for (TransactionInput input : tx.getInputs()) {
                input.sign(key, input.getConnectedOutput().getScriptPubKey());
            }

            // Finalize and return transaction hex
            promise.resolve(tx.bitcoinSerialize().toString());
        } catch (Exception e) {
            promise.reject("SIGNING_FAILED", e);
        }
    }

    private List<Utxo> fetchUtxos(String senderAddress) {
        // Implement network request to fetch UTXOs (e.g., using OkHttp)
        // This is a placeholder; replace with actual API call
        return new ArrayList<>();
    }

    private String convertToPsbt(Transaction tx) {
        // Placeholder: Implement PSBT conversion
        return "";
    }

    private Transaction convertFromPsbt(String psbtBase64) {
        // Placeholder: Implement PSBT parsing
        return new Transaction(PARAMS);
    }
}

File: ios/PayjoinModule.swift
import Foundation
import LibWally
import React

@objc(PayjoinModule)
class PayjoinModule: NSObject {
    @objc
    func createInitialPsbt(
        _ senderAddress: String,
        receiverAddress: String,
        amount: Int,
        feeRate: Int,
        resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        do {
            // Initialize network (testnet)
            let network = Network.testnet
            
            // Fetch UTXOs (simplified, assuming a helper function)
            let utxos = try fetchUtxos(senderAddress: senderAddress)
            
            // Create PSBT
            let psbt = try PSBT(network: network)
            var totalInput: UInt64 = 0
            
            // Add inputs
            for utxo in utxos {
                let txid = try Hash(fromHex: utxo.txid)
                let input = try PSBTInput(
                    txid: txid,
                    vout: UInt32(utxo.vout),
                    value: UInt64(utxo.amount),
                    scriptPubKey: try Script(hex: utxo.scriptPubKey)
                )
                try psbt.addInput(input)
                totalInput += UInt64(utxo.amount)
            }
            
            // Add receiver output
            let receiverScript = try Address(receiverAddress, network: network).scriptPubKey
            try psbt.addOutput(value: UInt64(amount), script: receiverScript)
            
            // Add change output
            let estimatedFee = UInt64(feeRate * 200 + 10000) // Max fee contribution
            let change = totalInput - UInt64(amount) - estimatedFee
            if change > 546 { // Dust threshold
                let changeScript = try Address(senderAddress, network: network).scriptPubKey
                try psbt.addOutput(value: change, script: changeScript)
            }
            
            // Return PSBT as base64
            resolve(psbt.description)
        } catch {
            reject("PSBT_CREATION_FAILED", error.localizedDescription, error)
        }
    }
    
    @objc
    func signAndFinalizePsbt(
        _ psbtBase64: String,
        privateKeyWif: String,
        resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        do {
            // Parse PSBT
            guard let psbt = try PSBT(psbtBase64) else {
                throw NSError(domain: "Invalid PSBT", code: -1, userInfo: nil)
            }
            
            // Sign with private key
            let key = try ECPrivateKey(wif: privateKeyWif, network: .testnet)
            try psbt.sign(key: key)
            
            // Finalize PSBT
            try psbt.finalize()
            
            // Extract transaction hex
            let tx = try psbt.extract()
            resolve(tx.hex)
        } catch {
            reject("SIGNING_FAILED", error.localizedDescription, error)
        }
    }
    
    // Helper function to fetch UTXOs (simplified)
    private func fetchUtxos(senderAddress: String) throws -> [Utxo] {
        // Implement network request to fetch UTXOs (e.g., using URLSession)
        // This is a placeholder; replace with actual API call
        return []
    }
}